rs/xnet/endpoint/src/lib.rs

```diff
@@ -153,90 +122,91 @@ impl XNetEndpoint {
 
         let metrics = Arc::new(XNetEndpointMetrics::new(metrics));
 
-        // The bounded channel for queuing requests between the HTTP server and the
-        // background worker.
-        //
-        // We use a background worker because building streams is CPU and memory
-        // bound and can take a few dozens of milliseconds, which is too long to execute
-        // on the event handling thread.
-        //
-        // We do not use [tokio::runtime::Handle::spawn_blocking] because it's designed
-        // for I/O bound tasks and can spawn extra threads when needed, which is
-        // not the best strategy for CPU-bound tasks.
-        //
-        // We use a crossbeam channel instead of [tokio::sync::mpsc] because we want the
-        // receiver to be blocking, and [tokio::sync::mpsc::Receiver::blocking_recv] is
-        // only available in tokio ≥ 0.3.
-        let (request_sender, request_receiver) =
-            crossbeam_channel::bounded(XNET_ENDPOINT_NUM_WORKER_THREADS);
-
-        let make_service = make_service_fn({
-            #[derive(Clone)]
-            struct Context {
-                log: ReplicaLogger,
-                request_sender: crossbeam_channel::Sender<WorkerMessage>,
-                metrics: Arc<XNetEndpointMetrics>,
-            }
+        // Spawn a request handler. We pass the certified stream store, which is
+        // currently realized by the state manager.
 
-            let ctx = Context {
-                log: log.clone(),
-                metrics: Arc::clone(&metrics),
-                request_sender: request_sender.clone(),
-            };
+        let make_service_closure = |address: SocketAddr| {
+            make_service_fn({
+                let base_url = Url::parse(&format!("http://{}/", address)).unwrap();
+                #[derive(Clone)]
+                struct Context {
+                    log: ReplicaLogger,
+                    semaphore: Arc<Semaphore>,
+                    metrics: Arc<XNetEndpointMetrics>,
+                }
 
-            fn ok<T>(t: T) -> Result<T, Infallible> {
-                Ok(t)
-            }
+                let ctx = Context {
+                    log: log.clone(),
+                    metrics: Arc::clone(&metrics),
+                    semaphore: Arc::new(Semaphore::new(XNET_ENDPOINT_NUM_WORKER_THREADS)),
```

Comment:
I believe the tokio runtime because like a standard threadpool. The number of blocking threads is unbounded - https://docs.rs/tokio/latest/tokio/runtime/index.html#lifetime-of-spawned-threads